### PR TITLE
raft: reduce allocation

### PR DESCRIPTION
**Pull request by BusyJay**

### CODE REVIEW

File: src/raft/raft.rs [hunk 1]
```diff
@@ -600,16 +599,23 @@ impl<T: Storage> Raft<T> {
...
-        let mut mis = Vec::with_capacity(self.prs.len());
-        for p in self.prs.values() {
-            mis.push(p.matched);
-        }
-        // reverse sort
-        mis.sort_by(|a, b| b.cmp(a));
-        let mci = mis[self.quorum() - 1];
-        let term = self.term;
-        self.raft_log.maybe_commit(mci, term)
+        let mci = if self.prs.len() <= 5 {
```
Comment (siddontang) [inline src/raft/raft.rs]: why < 5 here?

Comment (BusyJay) [reply src/raft/raft.rs]: So the matched IDs can be put in a [0; 5] array. Generally we only have 3 or 5 replicas.

Comment (overvenus) [reply src/raft/raft.rs]: How about `smallvec`[1] ?

[1] https://github.com/servo/rust-smallvec

Comment (BusyJay) [reply src/raft/raft.rs]: Why use it? I think the final code should look almost the same.

Comment (siddontang) [reply src/raft/raft.rs]: I think we should add a benchmark case here.

Comment (BusyJay) [reply src/raft/raft.rs]: The code will hardly change, an additional benchmark case seems redundant. My benchmark shows that use array is about 50% quicker.

When len is 3:
test bench_commit_array ... bench:          13 ns/iter (+/- 0)
test bench_commit_vec   ... bench:          28 ns/iter (+/- 0)

When len is 5:
test bench_commit_array ... bench:          20 ns/iter (+/- 0)
test bench_commit_vec   ... bench:          35 ns/iter (+/- 0)

File: src/raft/raft.rs [hunk 2]
```diff
@@ -600,16 +599,23 @@ impl<T: Storage> Raft<T> {
...
+            let mut mis = [0; 5];
+            for (pr, m) in self.prs.values().zip(&mut mis) {
+                *m = pr.matched;
+            }
+            mis[..self.prs.len()].sort_by(|a, b| b.cmp(a));
+            mis[self.quorum() - 1]
+        } else {
+            let mut mis = Vec::with_capacity(self.prs.len());
+            for p in self.prs.values() {
```
Comment (siddontang) [inline src/raft/raft.rs]: can we change vec to slice and unify the two similar codes?

File: src/raft/raw_node.rs [hunk 3]
```diff
@@ -126,9 +128,11 @@ impl Ready {
     ) -> Ready {
         let mut rd = Ready {
             entries: raft.raft_log.unstable_entries().unwrap_or(&[]).to_vec(),
-            messages: raft.msgs.drain(..).collect(),
             ..Default::default()
         };
+        if !raft.msgs.is_empty() {
+            mem::swap(&mut raft.msgs, &mut rd.messages);
```
Comment (siddontang) [inline src/raft/raw_node.rs]: any benchmark here?

Comment (BusyJay) [reply src/raft/raw_node.rs]: It's very obvious that `mem::swap` does not copy the heap values however `drain.collect` will allocate a new buffer then copy all the content from original.

For a 100 length u64 vector:
test bench_drain_collect ... bench:         105 ns/iter (+/- 2)
test bench_swap          ... bench:           2 ns/iter (+/- 0)

**Pull request merged**